File: sdk/android/story-sdk/src/main/kotlin/com/openstory/sdk/internal/network/OpenStoryApi.kt
package com.openstory.sdk.internal.network

import com.openstory.sdk.OpenStoryConfiguration
import java.io.IOException
import java.util.concurrent.TimeUnit
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody

internal class OpenStoryApi(
    configuration: OpenStoryConfiguration,
) {
    private val baseUrl = configuration.normalizedBaseUrl()

    private val client = OkHttpClient.Builder()
        .connectTimeout(configuration.connectTimeoutMillis, TimeUnit.MILLISECONDS)
        .readTimeout(configuration.readTimeoutMillis, TimeUnit.MILLISECONDS)
        .build()

    suspend fun fetchFeed(
        requestPayload: SdkFeedRequestPayload,
        staticToken: String,
    ): SdkFeedResponsePayload = withContext(Dispatchers.IO) {
        val body = SdkFeedJsonCodec.encodeRequest(requestPayload)
            .toRequestBody("application/json; charset=utf-8".toMediaType())

        val request = Request.Builder()
            .url("$baseUrl/v1/sdk/feed")
            .post(body)
            .header("Authorization", "Bearer $staticToken")
            .build()

        client.newCall(request).execute().use { response ->
            when {
                response.code == 401 || response.code == 403 -> {
                    throw OpenStoryAuthorizationException(response.code)
                }

                !response.isSuccessful -> {
                    throw IOException("Story feed request failed with HTTP ${response.code}.")
                }
            }

            val responseBody = response.body?.string()
                ?: throw IOException("Story feed response body is empty.")

            SdkFeedResponseNormalizer.normalizeLoopbackUrls(
                response = SdkFeedJsonCodec.decodeResponse(responseBody),
                baseUrl = baseUrl,
            )
        }
    }
}

internal class OpenStoryAuthorizationException(
    val statusCode: Int,
) : IOException("Story feed authorization failed with HTTP $statusCode.")

File: sdk/ios/Sources/OpenStorySDK/Internal/Network/OpenStoryAPI.swift
import Foundation

internal struct OpenStoryAuthorizationError: Error, Equatable {
    let statusCode: Int
}

internal final class OpenStoryAPI: @unchecked Sendable {
    private let baseURL: String
    private let session: URLSession
    private let encoder = JSONEncoder()
    private let decoder = JSONDecoder()

    init(
        configuration: OpenStoryConfiguration,
        session: URLSession? = nil
    ) {
        baseURL = configuration.normalizedBaseURL

        if let session {
            self.session = session
        } else {
            let sessionConfiguration = URLSessionConfiguration.ephemeral
            sessionConfiguration.timeoutIntervalForRequest = configuration.requestTimeoutInterval
            sessionConfiguration.timeoutIntervalForResource = configuration.resourceTimeoutInterval
            self.session = URLSession(configuration: sessionConfiguration)
        }
    }

    func fetchFeed(
        requestPayload: SdkFeedRequestPayload,
        staticToken: String
    ) async throws -> SdkFeedResponsePayload {
        guard let url = URL(string: "\(baseURL)/v1/sdk/feed") else {
            throw URLError(.badURL)
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json; charset=utf-8", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(staticToken)", forHTTPHeaderField: "Authorization")
        request.httpBody = try encoder.encode(requestPayload)

        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw URLError(.badServerResponse)
        }

        switch httpResponse.statusCode {
        case 401, 403:
            throw OpenStoryAuthorizationError(statusCode: httpResponse.statusCode)
        case 200 ..< 300:
            let decoded = try decoder.decode(SdkFeedResponsePayload.self, from: data)
            return SdkFeedResponseNormalizer.normalizeLoopbackURLs(
                response: decoded,
                baseURL: baseURL
            )
        default:
            throw NSError(
                domain: "OpenStoryAPI",
                code: httpResponse.statusCode,
                userInfo: [
                    NSLocalizedDescriptionKey: "Story feed request failed with HTTP \(httpResponse.statusCode)."
                ]
            )
        }
    }
}
